### PR TITLE
Delayed write ensemble change may cause dataloss

### PR DESCRIPTION
**Pull request by jvrao**
Descriptions of the changes in this PR:

The Original intent of this change is to do a best-effort ensemble change.
But this is not possible until the local metadata is completely immutable.
Until the feature "Make LedgerMetadata Immutable #610" Is complete we will use
handleBookieFailure() to handle delayed writes as regular bookie failures.

Signed-off-by: Venkateswararao Jujjuri (JV) <vjujjuri@salesforce.com>

Master Issue: #1591
Relate Issue: #1395 

> ---
> In order to uphold a high standard for quality for code contributions, Apache BookKeeper runs various precommit
> checks for pull requests. A pull request can only be merged when it passes precommit checks. However running all
> the precommit checks can take a long time, some trivial changes don't need to run all the precommit checks. You
> can check following list to skip the tests that don't need to run for your pull request. Leave them unchecked if
> you are not sure, committers will help you:
>
> - [ ] [skip bookkeeper-server bookie tests]: skip testing `org.apache.bookkeeper.bookie` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server client tests]: skip testing `org.apache.bookkeeper.client` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server replication tests]: skip testing `org.apache.bookkeeper.replication` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server tls tests]: skip testing `org.apache.bookkeeper.tls` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server remaining tests]: skip testing all other tests in bookkeeper-server module.
> - [ ] [skip integration tests]: skip docker based integration tests. if you make java code changes, you shouldn't skip integration tests.
> - [ ] [skip build java8]: skip build on java8. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> - [ ] [skip build java9]: skip build on java9. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> ---

> ---
> Be sure to do all of the following to help us incorporate your contribution
> quickly and easily:
>
> If this PR is a BookKeeper Proposal (BP):
>
> - [ ] Make sure the PR title is formatted like:
>     `<BP-#>: Description of bookkeeper proposal`
>     `e.g. BP-1: 64 bits ledger is support`
> - [ ] Attach the master issue link in the description of this PR.
> - [ ] Attach the google doc link if the BP is written in Google Doc.
>
> Otherwise:
> 
> - [ ] Make sure the PR title is formatted like:
>     `<Issue #>: Description of pull request`
>     `e.g. Issue 123: Description ...`
> - [ ] Make sure tests pass via `mvn clean apache-rat:check install spotbugs:check`.
> - [ ] Replace `<Issue #>` in the title with the actual Issue number.
> 
> ---

### CODE REVIEW

**Comment by jvrao** [comment]
Looks like there is some issue with DeferredSyncTest

Seeing a bunch of write failures with this error, and with the latest fix, it is causing ensemble change and causing the test failure 

'''
2018-08-09 20:20:28,196 - ERROR - [Builder-OrderedExecutor-1-0:DigestManager@170] - Data received is smaller than the minimum for this digest type.  Either the packet it corrupt, or the wrong digest is configured.  Digest type: org.apache.bookkeeper.proto.checksum.CRC32DigestManager, Packet Length: 0
'''

@eolivelli  @sijie

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java [hunk 1]
```diff
@@ -1867,48 +1867,59 @@ EnsembleInfo replaceBookieInMetadata(final Map<Integer, BookieSocketAddress> fai
...
-                            getId(), curNumEnsembleChanges, ensembleInfo);
-                }
-                writeLedgerConfig(new ChangeEnsembleCb(ensembleInfo, curBlockAddCompletions,
-                        curNumEnsembleChanges, false));
-            } catch (BKException.BKNotEnoughBookiesException e) {
-                LOG.error("Could not get additional bookie to remake ensemble: {}", ledgerId);
-            }
-            delayedWriteFailedBookies.clear();
-        }
+        final Map<Integer, BookieSocketAddress> CopyDelayedWriteFailedBookies =
```
Comment (ivankelly) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java]: nit: variable names should start with a lower case

Comment (jvrao) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java]: Thanks; this is causing some test failures. Looking into it.

**Comment by jvrao** [comment]
@sijie @ivankelly added more cleanup.

**Comment by ivankelly** [comment]
@jvrao the two dependent changes are merged, so I merged master back into your branch.

**Comment by ivankelly** [comment]
rerun integration tests

**Comment by ivankelly** [comment]
Don't close yet. I want to submit a test for this.

**Comment by sijie** [comment]
@ivankelly sorry I didn't see any comments for tests. since this is already merged, it will be much clearer if you send out a separated PR for adding the tests and link back to this PR.

**Comment by sijie** [comment]
or I think you should reopen #1591 (which is an issue), not #1592 (which is a PR).

**Comment by ivankelly** [comment]
Ah, sorry, thought this was the issue.

**Pull request closed without merging**